### PR TITLE
integrate support for mfu tracking, including hardware-specific peak TFLOPs detection

### PR DESCRIPTION
**Pull request by Luodian**
# Add MFU Tracking and Hardware-Specific Peak TFLOPs Detection

This PR integrates Model FLOPs Utilization (MFU) tracking into the training pipeline, enabling real-time monitoring of GPU compute efficiency.

The implementation includes:

1. Automatic hardware detection for various GPU models (MI300X, H100/H800/H200, A100/A800, L40/L20, H20, Ascend 910B, etc.).
2. FLOPs calculation for the Qwen2/2.5/3 architecture, and training metrics including tokens/sec, actual TFLOPs, and MFU percentage.
3. Accept manual specification via the `--peak_device_tflops` argument. 

Also fixed some small issues:

1. Additional improvements include gradient accumulation (`--gradient_accumulation_steps`), better OOM error handling.
2. Checkpoint saving with proper CUDA synchronization and memory management, model parameter counting, and automatic final checkpoint save at training completion.

### CODE REVIEW

**Comment by Luodian** [comment]
PixPin_2025-10-27_00-29-49

**Comment by Luodian** [comment]
training script:

```
#!/bin/bash
# Copyright 2025 Bytedance Ltd. and/or its affiliates.
# SPDX-License-Identifier: Apache-2.0

# Simple 8 GPU training test with MFU calculation

export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
export WANDB_MODE=offline  # Run wandb offline for testing
export PYTHONPATH=/opt/tiger/Bagel:$PYTHONPATH
export HF_HUB_ENABLE_HF_TRANSFER=0  # Disable hf_transfer

cd /opt/tiger/Bagel
source .venv/bin/activate

python -m torch.distributed.run \
  --nnodes=1 \
  --node_rank=0 \
  --nproc_per_node=8 \
  --master_addr=localhost \
  --master_port=39500 \
  train/pretrain_unified_navit.py \
  --dataset_config_file ./data/configs/example.yaml \
  --model_path /opt/tiger/BAGEL-7B-MoT \
  --resume-from /opt/tiger/BAGEL-7B-MoT \
  --layer_module Qwen2MoTDecoderLayer \
  --max_latent_size 64 \
  --finetune_from_hf True \
  --auto_resume True \
  --visual_und True \
  --visual_gen True \
  --resume-model-only True \
  --finetune-from-ema True \
  --log_every 1 \
  --lr 4e-6 \
  --num_worker 1 \
  --expected_num_tokens 10240 \
  --max_num_tokens 11520 \
  --max_num_tokens_per_sample 10240 \
  --num_shard 8 \
  --wandb_project "BAGEL-i2t-finetune" \
  --wandb_name "BAGEL-i2t-finetune"
```

**Comment by Andy1621** [comment]
LGTM
